src/NumericalAlgorithms/Spectral/SwshDerivatives.cpp

```diff
@@ -0,0 +1,117 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "NumericalAlgorithms/Spectral/SwshDerivatives.hpp"
+
+#include "DataStructures/ComplexDataVector.hpp"  // IWYU pragma: keep
+#include "DataStructures/ComplexDiagonalModalOperator.hpp"
+#include "DataStructures/ComplexModalVector.hpp"
+#include "DataStructures/SpinWeighted.hpp"  // IWYU pragma: keep
+#include "DataStructures/TempBuffer.hpp"
+#include "NumericalAlgorithms/Spectral/ComplexDataView.hpp"
+#include "NumericalAlgorithms/Spectral/SwshCollocation.hpp"
+#include "NumericalAlgorithms/Spectral/SwshTransform.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/StaticCache.hpp"
+
+// IWYU pragma: no_forward_declare SpinWeighted
+
+namespace Spectral {
+namespace Swsh {
+namespace detail {
+
+template <typename DerivativeKind, int Spin>
+ComplexDiagonalModalOperator derivative_factors(size_t l_max) noexcept {
+  ComplexDiagonalModalOperator derivative_factors{
+      size_of_libsharp_coefficient_vector(l_max)};
+  for (const auto& mode : cached_coefficients_metadata(l_max)) {
+    // note that the libsharp transform data is stored as a pair of complex
+    // vectors: one for the transform of the real part and one for the transform
+    // of the imaginary part of the collocation data.
+    derivative_factors[mode.transform_of_real_part_offset] =
+        (detail::derivative_factor<DerivativeKind>(mode.l, Spin));
+    derivative_factors[mode.transform_of_imag_part_offset] =
+        (detail::derivative_factor<DerivativeKind>(mode.l, Spin));
+  }
+  // apply the sign change as appropriate due to the adjusted spin
+  for (size_t i = 0; i < 2; i++) {
```

Comment:
why `2` as the upper bound? Could you either add a comment in the code or come up with a more descriptive name than `i` if there is one?